File: Videos1_20/CSharpcode/Toolbar.cs
using System.Windows;
using System.Windows.Controls;

namespace CSharpCode {
    public class Toolbar : UserControl {
        private Button helloButton;
        private Button goodbyeButton;
        private StringListener? textListener;

        public Toolbar() {
            StackPanel panel = new StackPanel { Orientation = Orientation.Horizontal, Margin = new Thickness(5) };
            
            helloButton = new Button { Content = "Hello", Width = 60, Margin = new Thickness(5) };
            goodbyeButton = new Button { Content = "Goodbye", Width = 60, Margin = new Thickness(5) };
            
            helloButton.Click += (s, e) => {
                if (textListener != null) textListener.textEmitted("Hello\n");
            };
            
            goodbyeButton.Click += (s, e) => {
                if (textListener != null) textListener.textEmitted("Goodbye\n");
            };
            
            panel.Children.Add(helloButton);
            panel.Children.Add(goodbyeButton);
            Content = panel;
        }

        public void setStringListener(StringListener listener) {
            this.textListener = listener;
        }
    }
}

File: Videos1_20/PythonCode/Toolbar.py
import tkinter as tk

class Toolbar(tk.Frame):
    def __init__(self, parent):
        super().__init__(parent, relief=tk.GROOVE, borderwidth=2)
        self.textListener = None
        
        self.helloButton = tk.Button(self, text="Hello", command=lambda: self.actionPerformed("hello"))
        self.helloButton.pack(side=tk.LEFT, padx=5, pady=5)
        
        self.goodbyeButton = tk.Button(self, text="Goodbye", command=lambda: self.actionPerformed("goodbye"))
        self.goodbyeButton.pack(side=tk.LEFT, padx=5, pady=5)

    def setStringListener(self, listener):
        self.textListener = listener

    def actionPerformed(self, action):
        if action == "hello" and self.textListener:
            self.textListener.textEmitted("Hello\n")
        elif action == "goodbye" and self.textListener:
            self.textListener.textEmitted("Goodbye\n")